File: frontend/src/types/sse.ts
/**
 * TS mirror of the backend SSE union (`app/schemas/sse.py`) — the frontend half of the
 * drift gate. Keep field-for-field in sync; the shared fixture
 * (`contracts/sse_events.fixture.json`) is asserted against BOTH sides.
 *
 * Design: the answer streams ONLY on `token` events; all validated control metadata rides
 * the single trailing `final` event.
 */

export type StatusStage = "retrieving" | "looking_up" | "generating" | "waiting";

export interface StatusEvent {
  type: "status";
  stage: StatusStage;
}

export interface TokenEvent {
  type: "token";
  text: string;
}

export interface Citation {
  index: number;
  source_id: string;
  title: string;
  page_number?: number; // file sources -> "from <file>, p.<n>"
  source_url?: string; // crawled pages -> clickable link
}

export interface CitationEvent {
  type: "citation";
  citation: Citation;
}

export interface Tag {
  name: string;
  status: "approved" | "pending";
}

export interface FinalEvent {
  type: "final";
  answer_complete: boolean;
  detected_language: string; // BCP-47
  tags: Tag[];
  retrieval_hits: number;
  escalate: boolean;
  escalation_reason?: string; // canonical enum value
  advisory_confidence?: number; // observability only, NOT the escalation trigger
}

export interface ErrorEvent {
  type: "error";
  code: string;
  message: string;
}

export interface DoneEvent {
  type: "done";
  turn_id: string;
  ticket_state: string;
}

export type SSEEvent =
  | StatusEvent
  | TokenEvent
  | CitationEvent
  | FinalEvent
  | ErrorEvent
  | DoneEvent;

const STAGES: readonly StatusStage[] = ["retrieving", "looking_up", "generating", "waiting"];

function isObject(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null;
}

/**
 * Runtime validator + type-narrower. Throws on an unknown `type` or a malformed payload —
 * this is what fails CI if the backend union drifts from this mirror.
 */
export function parseSSEEvent(data: unknown): SSEEvent {
  if (!isObject(data) || typeof data.type !== "string") {
    throw new Error(`SSE event is not an object with a string 'type': ${JSON.stringify(data)}`);
  }
  switch (data.type) {
    case "status":
      if (!STAGES.includes(data.stage as StatusStage)) {
        throw new Error(`invalid status stage: ${String(data.stage)}`);
      }
      return { type: "status", stage: data.stage as StatusStage };
    case "token":
      if (typeof data.text !== "string") throw new Error("token.text must be a string");
      return { type: "token", text: data.text };
    case "citation": {
      const c = data.citation;
      if (!isObject(c) || typeof c.index !== "number" || typeof c.source_id !== "string" || typeof c.title !== "string") {
        throw new Error("malformed citation event");
      }
      return {
        type: "citation",
        citation: {
          index: c.index,
          source_id: c.source_id,
          title: c.title,
          page_number: typeof c.page_number === "number" ? c.page_number : undefined,
          source_url: typeof c.source_url === "string" ? c.source_url : undefined,
        },
      };
    }
    case "final":
      if (typeof data.detected_language !== "string") {
        throw new Error("final.detected_language must be a string");
      }
      return {
        type: "final",
        answer_complete: Boolean(data.answer_complete),
        detected_language: data.detected_language,
        tags: Array.isArray(data.tags) ? (data.tags as Tag[]) : [],
        retrieval_hits: typeof data.retrieval_hits === "number" ? data.retrieval_hits : 0,
        escalate: Boolean(data.escalate),
        escalation_reason:
          typeof data.escalation_reason === "string" ? data.escalation_reason : undefined,
        advisory_confidence:
          typeof data.advisory_confidence === "number" ? data.advisory_confidence : undefined,
      };
    case "error":
      if (typeof data.code !== "string" || typeof data.message !== "string") {
        throw new Error("malformed error event");
      }
      return { type: "error", code: data.code, message: data.message };
    case "done":
      if (typeof data.turn_id !== "string" || typeof data.ticket_state !== "string") {
        throw new Error("malformed done event");
      }
      return { type: "done", turn_id: data.turn_id, ticket_state: data.ticket_state };
    default:
      throw new Error(`unknown SSE event type: ${data.type}`);
  }
}

File: app/schemas/sse.py
"""The SSE event protocol — the single source of truth for the streamed chat turn.

OpenAPI codegen cannot describe an ``text/event-stream``, so this discriminated union is
authored once here (Pydantic) and mirrored to a TS type in ``frontend/src/types/sse.ts``.
A shared fixture (``contracts/sse_events.fixture.json``) is asserted by both a backend test
(``tests/test_sse_contract.py``) and a frontend test — the drift gate (audit [A?]/[IMP-FE-5]).

Key design point [IMP-ENG-1]: the natural-language answer streams ONLY on ``token`` events;
all validated control metadata (tags, escalate, detected_language, answer_complete, …) rides
the single trailing ``final`` event, so streaming and JSON-validation never fight.
"""

from __future__ import annotations

import json
from typing import Annotated, Literal, Union

from pydantic import BaseModel, Field, TypeAdapter

from app.domain.escalation.reasons import EscalationReason


class StatusEvent(BaseModel):
    """Progress affordance during the pre-first-token tool phase. The widget maps ``stage``
    to a pre-translated chrome string, so filler is localised without server-side language
    knowledge [IMP-ENG-7]."""

    type: Literal["status"] = "status"
    stage: Literal["retrieving", "looking_up", "generating", "waiting"]


class TokenEvent(BaseModel):
    type: Literal["token"] = "token"
    text: str


class Citation(BaseModel):
    index: int
    source_id: str
    title: str
    page_number: int | None = None  # [A13] file sources render "from <file>, p.<n>"
    source_url: str | None = None    # [A13] crawled pages render a clickable link


class CitationEvent(BaseModel):
    type: Literal["citation"] = "citation"
    citation: Citation


class Tag(BaseModel):
    """Tags are {name, status} objects, never plain strings (PRD §4.2.3)."""

    name: str
    status: Literal["approved", "pending"]


class FinalEvent(BaseModel):
    """The validated per-turn control envelope. Emitted once, after the answer tokens."""

    type: Literal["final"] = "final"
    answer_complete: bool = False  # [A5] gates the closing question / auto-resolve path
    detected_language: str  # BCP-47, e.g. "en", "es", "hi"
    tags: list[Tag] = Field(default_factory=list)
    retrieval_hits: int = 0
    escalate: bool = False
    escalation_reason: EscalationReason | None = None
    # Observability only — NEVER the escalation trigger (that is retrieval-grounding-based).
    advisory_confidence: float | None = None


class ErrorEvent(BaseModel):
    type: Literal["error"] = "error"
    code: str
    message: str


class DoneEvent(BaseModel):
    type: Literal["done"] = "done"
    turn_id: str
    ticket_state: str


SSEEvent = Annotated[
    Union[StatusEvent, TokenEvent, CitationEvent, FinalEvent, ErrorEvent, DoneEvent],
    Field(discriminator="type"),
]

_ADAPTER: TypeAdapter[SSEEvent] = TypeAdapter(SSEEvent)


def parse_event(data: dict | str) -> SSEEvent:
    """Validate an inbound event payload against the union (raises on drift)."""
    if isinstance(data, (bytes, bytearray, str)):
        return _ADAPTER.validate_json(data)
    return _ADAPTER.validate_python(data)


def to_sse_frame(event: SSEEvent, seq: int | None = None) -> str:
    """Render an event as a wire SSE frame: ``event:``/``id:``/``data:`` lines.

    ``id:`` is a monotonic sequence carried at the transport layer only (kept trivially, per
    trim [T7]); reconnect relies on ``client_msg_id`` idempotent whole-turn replay, not a
    per-event replay buffer.
    """
    payload = event.model_dump(mode="json", exclude_none=True)
    lines = [f"event: {payload['type']}"]
    if seq is not None:
        lines.append(f"id: {seq}")
    lines.append(f"data: {json.dumps(payload, separators=(',', ':'))}")
    return "\n".join(lines) + "\n\n"
